### PR TITLE
fix(es/compat): Handle private names from class properties pass

### LINKED ISSUE #7561: Private property after an optional chain transforms into nonsense code
### Describe the bug

I think this also broke when we implemented the new transform. Currently, if you have a private property access after an optional chain, the chain will run. The `_memo == null` conditional check happens _before_ the assignment of `_memo`.

### Input code

```typescript
class Foo {
  #x;

  test() {
    this?.y.#x
  }
}
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "ecmascript",
      "jsx": false
    },
    "target": "es2021",
    "loose": false,
    "minify": {
      "compress": false,
      "mangle": false
    }
  },
  "module": {
    "type": "commonjs"
  },
  "minify": false,
  "isModule": false
}
```

### Playground link

https://play.swc.rs/?version=1.3.66&code=H4sIAAAAAAAAA0vOSSwuVnDLz1eo5lJQUK6w5gJSJanFJRqaYBEgJyOz2F6vUk%2B5Asit5aoFAHwm0I8yAAAA&config=H4sIAAAAAAAAA1VPOw7DIAzdcwrkuUObMXfoISzqRET8hKlUFOXuBQJps%2Fn95W0QAlaWMIktnxl4DEzhxJnhZCN%2BMgMkDbIMyke4dXXlIs2omSq1HwpEDAvFmuLxPj5aArRzTD3ROKOsmtP%2FpnTGB2K%2BGosV7aLpuji0VTDu9a5i%2ByUmX1BpM86uDD9nXzzbQfGzx4%2Fu%2FQutaQ0NHQEAAA%3D%3D

### Expected behavior

The memo is assigned before the conditional, and the private access is scoped to the conditional's alternate branch:

```js
// …
var _x = new WeakMap();
class Foo {
    test() {
        (_this = this) === null || _this === void 0
          ? void 0
          : _class_private_field_get(_this.y , _x);
    }
    // …
}
```

### Actual behavior

The chain is nonsense:

```js
// …
var _x = new WeakMap();
class Foo {
    test() {
        var _this_y, _this_y1;
        _this_y === null || _this_y === void 0
            ? void 0
            : _class_private_field_get(_this_y1 = this?.y, _x);
    }
    // …
}
```

### Version

1.3.66

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kdy1**
**Description:**

**Related issue:**

 - Closes #7561
 - https://github.com/vercel/next.js/issues/56612

### CODE REVIEW

**Review by kdy1 — COMMENTED** [review]
swc-bump:
 - swc_ecma_compat_es2020
 - swc_ecma_compat_es2022

File: crates/swc/tests/fixture/issues-7xxx/7561/output/1.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+var _class_private_field_get = require("@swc/helpers/_/_class_private_field_get");
+var _class_private_field_init = require("@swc/helpers/_/_class_private_field_init");
+var _x = /*#__PURE__*/ new WeakMap();
+class Foo {
+    test() {
+        var _this_y, _this_y1;
+        _this_y === null || _this_y === void 0 ? void 0 : _class_private_field_get._(_this_y1 = this?.y, _x);
```
Comment (jridgewell) [inline crates/swc/tests/fixture/issues-7xxx/7561/output/1.js:7]: This is still broken. What's `_this_y` isn't assigned, `this?.y` still remains in the call args.

**Review by jridgewell — CHANGES_REQUESTED** [review]

File: crates/swc/tests/fixture/issues-7xxx/7561/output/1.js (hunk 1 above)
Comment (kdy1) [reply crates/swc/tests/fixture/issues-7xxx/7561/output/1.js]: It was a problem with the target ES version.
I added tests to show the result per the target ES version

File: crates/swc/tests/fixture/issues-7xxx/7561/es2019/output/1.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+var _class_private_field_get = require("@swc/helpers/_/_class_private_field_get");
+var _class_private_field_init = require("@swc/helpers/_/_class_private_field_init");
+var _x = /*#__PURE__*/ new WeakMap();
+class Foo {
+    test() {
+        var _this;
```
Comment (kdy1) [inline crates/swc/tests/fixture/issues-7xxx/7561/es2019/output/1.js:6]: This does not look super efficient, though. I'll fix it with another PR

Comment (kdy1) [reply crates/swc/tests/fixture/issues-7xxx/7561/es2019/output/1.js:6]: Nevermind. It's used

**Review by jridgewell — CHANGES_REQUESTED** [review]
All of the outputs only show the assignment happening in the call arg, **after** the conditional has already early short-circuited:

```js
_this_y === null || _this_y === void 0 ? void 0 : _class_private_field_get._(_this_y1 = (_this = this) === null || _this === void 0 ? void 0 : _this.y, _x);
```

The private names transform itself needs to handle optional chaining. We did the same thing in Babel, essentially reimplmementing it in both trasnformers.

File: crates/swc_ecma_transforms_compat/tests/optional-chaining/issue-7561/output.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+var _x = /*#__PURE__*/ new WeakMap();
+class Foo {
+    test() {
+        var _this;
+        var _this_y, _this_y1;
+        _this_y === null || _this_y === void 0 ? void 0 : _class_private_field_get(_this_y1 = (_this = this) === null || _this === void 0 ? void 0 : _this.y, _x);
```
Comment (jridgewell) [inline crates/swc_ecma_transforms_compat/tests/optional-chaining/issue-7561/output.js:6]: I'm not sure where our miscommunication is coming from, but this is still incorrect.
```diff
- _this_y === null || _this_y === void 0 ? void 0 : _class_private_field_get(_this_y1 = (_this = this) === null || _this === void 0 ? void 0 : _this.y, _x);
+ (_this = this) === null || _this === void 0 ? void 0 : _class_private_field_get(_this.y, _x);
```

**Review by jridgewell — CHANGES_REQUESTED** [review]

File: crates/swc/tests/fixture/issues-7xxx/7561/es2019/output/1.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+var _class_private_field_get = require("@swc/helpers/_/_class_private_field_get");
+var _class_private_field_init = require("@swc/helpers/_/_class_private_field_init");
+var _x = /*#__PURE__*/ new WeakMap();
+class Foo {
+    test() {
+        var _this, _this_y;
```
Comment (kdy1) [inline crates/swc/tests/fixture/issues-7xxx/7561/es2019/output/1.js:6]: This is not ideal

**Review by kdy1 — COMMENTED** [review]
swc-bump:
 - swc_ecma_compat_es2020
 - swc_ecma_compat_es2022

File: crates/swc/tests/fixture/issues-7xxx/7561/es2019/output/1.js (hunk 4 above)
Comment (kdy1) [reply crates/swc/tests/fixture/issues-7xxx/7561/es2019/output/1.js:6]: @jridgewell  But let's fix correctness issues first.

File: crates/swc/tests/tsc-references/privateNameFieldCallExpression.1.normal.js [hunk 5]
```diff
@@ -8,7 +8,7 @@ class A {
         var _class_private_field_get1;
         var _this_getInstance;
         _class_private_field_get(this, _fieldFunc).call(this);
-        (_class_private_field_get1 = _class_private_field_get(this, _fieldFunc)) === null || _class_private_field_get1 === void 0 ? void 0 : _class_private_field_get1.call(this);
+        (_class_private_field_get1 = _class_private_field_get(this, _fieldFunc)) === null || _class_private_field_get1 === void 0 ? void 0 : _class_private_field_get1();
```
Comment (kdy1) [inline crates/swc/tests/tsc-references/privateNameFieldCallExpression.1.normal.js:11]: This is wrong

**Review by kdy1 — COMMENTED** [review]
swc-bump:
 - swc_ecma_parser

File: crates/swc/tests/tsc-references/privateNameFieldCallExpression.1.normal.js [hunk 6]
```diff
@@ -5,10 +5,9 @@ import { _ as _class_private_field_set } from "@swc/helpers/_/_class_private_fie
 var _fieldFunc = /*#__PURE__*/ new WeakMap(), _fieldFunc2 = /*#__PURE__*/ new WeakMap();
 class A {
     test() {
-        var _class_private_field_get1;
-        var _this_getInstance;
+        var _this, _this1, _ref, _this_getInstance;
```
Comment (kdy1) [inline crates/swc/tests/tsc-references/privateNameFieldCallExpression.1.normal.js:8]: Previous output seems wrong

**Review by jridgewell — APPROVED** [review]

**Pull request merged**